File: Dice/Dice/ShakeObserver.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;

using System.Linq;
using Microsoft.Devices.Sensors;
using Microsoft.Phone.Reactive;

namespace Dice
{
    public class ShakeObserver
    {
        const double MinimumOffset = 1.44;
        const double TimeThreshold = 200;
        public static IObservable<IEvent<AccelerometerReadingEventArgs>> 
            GetObserver(Accelerometer accel)
        {
            var readingChangedObservable =
                Observable.FromEvent<AccelerometerReadingEventArgs>
                (accel, "ReadingChanged");

            var query =  
                from interval in 
                  (from startEvent in readingChangedObservable
                   where (startEvent.EventArgs.X * startEvent.EventArgs.X + 
                        startEvent.EventArgs.Y * startEvent.EventArgs.Y) >
                        MinimumOffset
                   select startEvent).TimeInterval()
                where interval.Interval.TotalMilliseconds < TimeThreshold 
                select interval.Value;

            return query;
        }
    }
}

File: Dice/Gestures/ShakeObserver.vb
Imports Microsoft.Phone.Reactive
Imports Microsoft.Devices.Sensors

Public Module ShakeObserver
    Const MinimumOffset = 1.44
    Const TimeThreshold = 200

    Public Function GetShakeObserver(ByVal accel As Accelerometer) As IObservable(Of Unit)

        Dim Movements = From start In Observable.FromEvent(Of AccelerometerReadingEventArgs)(accel, "ReadingChanged")
                        Where (start.EventArgs.X ^ 2 + start.EventArgs.Y ^ 2) > MinimumOffset

        Dim shake = From movement In Movements.TimeInterval
                    Where movement.Interval.TotalMilliseconds < TimeThreshold
                    Select New Unit()

        Return shake.Throttle(TimeSpan.FromSeconds(1.5)).
            ObserveOnDispatcher()

    End Function

End Module